File: packages/ts/custom/src/tool-call.ts
import { redactStatusMessage } from "./redaction";
import type { ToolCallOptions } from "./types";
import { debug } from "./transport";

/** @internal Wire-format payload for a tool call. */
export type ToolCallPayload = {
  type: "tool_call";
  run_id: string;
  tool_call_id: string;
  tool_name: string;
  start_time: string;
  end_time: string;
  status_code: number;
  status_message?: string;
  args?: string;
  result?: string;
};

/**
 * A tool call represents a single tool/function invocation within a
 * {@link Run}.
 *
 * Tool calls are created via {@link Run.toolCall | run.toolCall()} and
 * are automatically batched with the parent run when it is sent.
 *
 * A tool name must be set (either in the constructor or via
 * {@link ToolCall.name | .name()}) before calling {@link ToolCall.end | .end()}.
 *
 * @example
 * ```ts
 * const tc = r.toolCall("get_weather");
 * tc.args({ city: "San Francisco" });
 * tc.result({ temp: 72, unit: "F" });
 * tc.end();
 * ```
 */
export class ToolCall {
  private _runId: string;
  private _toolCallId: string;
  private _startTime: string;
  private _endTime: string | null = null;

  private _name: string | null = null;

  private _statusCode = 0;
  private _statusMessage: string | null = null;

  private _args: string | null = null;
  private _result: string | null = null;

  private _ended = false;

  constructor(runId: string, nameOrOptions?: string | ToolCallOptions) {
    this._runId = runId;

    if (typeof nameOrOptions === "string") {
      this._name = nameOrOptions;
      this._toolCallId = crypto.randomUUID();
      this._startTime = new Date().toISOString();
    } else {
      this._toolCallId = nameOrOptions?.toolCallId ?? crypto.randomUUID();
      this._startTime = (nameOrOptions?.startTime ?? new Date()).toISOString();
      if (nameOrOptions?.name) this._name = nameOrOptions.name;
    }

    debug("ToolCall created", { toolCallId: this._toolCallId, runId });
  }

  /** Whether this tool call has been finalized via `.end()` or `.error()`. */
  get ended(): boolean {
    return this._ended;
  }

  /**
   * Set the tool name (e.g. `"search"`, `"get_weather"`).
   * Must be set before {@link ToolCall.end | .end()} — either here or via
   * the constructor.
   *
   * @returns `this` for chaining.
   */
  name(toolName: string): this {
    this._name = toolName;
    return this;
  }

  /**
   * Set the arguments passed to the tool. Objects are auto-serialized to JSON.
   *
   * @param value - A JSON string or a plain object.
   * @returns `this` for chaining.
   */
  args(value: string | Record<string, unknown>): this {
    this._args = typeof value === "string" ? value : JSON.stringify(value);
    return this;
  }

  /**
   * Set the return value from the tool. Objects are auto-serialized to JSON.
   *
   * @param value - A JSON string or a plain object.
   * @returns `this` for chaining.
   */
  result(value: string | Record<string, unknown>): this {
    this._result = typeof value === "string" ? value : JSON.stringify(value);
    return this;
  }

  /**
   * Override the tool call's end time. By default, the end time is captured
   * automatically when `.end()` or `.error()` is called.
   *
   * @returns `this` for chaining.
   */
  endTime(date: Date): this {
    this._endTime = date.toISOString();
    return this;
  }

  /**
   * Set the outcome status code and an optional human-readable message.
   *
   * @param code - `0` for success, `2` for error.
   * @param message - Optional status message.
   * @returns `this` for chaining.
   */
  status(code: number, message?: string): this {
    this._statusCode = code;
    if (message !== undefined) this._statusMessage = redactStatusMessage(message);
    return this;
  }

  /**
   * Mark this tool call as errored (status code `2`).
   *
   * @param message - Optional error message.
   * @throws If the tool call has already been ended.
   */
  error(message = ""): void {
    if (this._ended) throw new Error("[TCC] ToolCall already ended");
    this._statusCode = 2;
    if (message) this._statusMessage = redactStatusMessage(message);
    this._ended = true;
    this._endTime ??= new Date().toISOString();
    debug("ToolCall error", { toolCallId: this._toolCallId, message });
  }

  /**
   * Finalize this tool call. A tool name must have been set (via the
   * constructor or {@link ToolCall.name | .name()}) before calling this.
   *
   * @throws If the tool call has already been ended.
   * @throws If no tool name was set.
   */
  end(): void {
    if (this._ended) throw new Error("[TCC] ToolCall already ended");
    if (this._name === null) {
      throw new Error(
        "[TCC] ToolCall requires a name. Call .name() or pass it to run.toolCall('name') before .end()"
      );
    }
    this._ended = true;
    this._endTime ??= new Date().toISOString();
    debug("ToolCall ended", { toolCallId: this._toolCallId });
  }

  /** @internal */
  _toPayload(): ToolCallPayload {
    const payload: ToolCallPayload = {
      type: "tool_call",
      run_id: this._runId,
      tool_call_id: this._toolCallId,
      tool_name: this._name ?? "unknown",
      start_time: this._startTime,
      end_time: this._endTime ?? new Date().toISOString(),
      status_code: this._statusCode,
    };

    if (this._statusMessage !== null)
      payload.status_message = this._statusMessage;
    if (this._args !== null) payload.args = this._args;
    if (this._result !== null) payload.result = this._result;

    return payload;
  }
}

File: packages/python/contextcompany/tool_call.py
import json
import uuid
from typing import Any, Dict, Optional, Union

from ._utils import _now_iso, _debug, _send_payload
from .redaction import redact_status_message


class ToolCall:
    def __init__(
        self,
        run_id: str,
        tool_call_id: Optional[str] = None,
        tool_name: Optional[str] = None,
        api_key: Optional[str] = None,
        tcc_url: Optional[str] = None,
    ) -> None:
        self._run_id = run_id
        self._tool_call_id = tool_call_id or str(uuid.uuid4())
        self._api_key = api_key
        self._tcc_url = tcc_url

        self._start_time: str = _now_iso()

        self._name: Optional[str] = tool_name

        self._status_code: int = 0
        self._status_message: Optional[str] = None

        self._args: Optional[str] = None
        self._result: Optional[str] = None

        self._ended = False

        _debug("ToolCall created")
        _debug("tool_call_id:", self._tool_call_id)
        _debug("run_id:", self._run_id)
        _debug("start_time:", self._start_time)

    def name(self, tool_name: str) -> "ToolCall":
        self._name = tool_name
        _debug("ToolCall name set:", tool_name)
        return self

    def args(self, value: Union[str, Dict[str, Any]]) -> "ToolCall":
        self._args = value if isinstance(value, str) else json.dumps(value)
        _debug("ToolCall args set:", self._args[:200] if len(self._args) > 200 else self._args)
        return self

    def result(self, value: Union[str, Dict[str, Any]]) -> "ToolCall":
        self._result = value if isinstance(value, str) else json.dumps(value)
        _debug("ToolCall result set:", self._result[:200] if len(self._result) > 200 else self._result)
        return self

    def status(self, code: int, message: Optional[str] = None) -> "ToolCall":
        self._status_code = code
        if message is not None:
            self._status_message = redact_status_message(message)
        _debug("ToolCall status set:", code, message)
        return self

    def error(self, status_message: str = "") -> None:
        if self._ended:
            raise RuntimeError("[TCC] ToolCall has already ended")

        _debug("ToolCall error:", status_message)
        self._status_code = 2
        if status_message:
            self._status_message = redact_status_message(status_message)
        self._ended = True

        payload = self._build_payload()
        _send_payload(payload, "tool_call", api_key=self._api_key, tcc_url=self._tcc_url)

    def end(self) -> None:
        if self._ended:
            raise RuntimeError("[TCC] ToolCall has already ended")

        if self._name is None:
            raise ValueError(
                "[TCC] Cannot end tool call: name is required. "
                "Call tc.name(...) or pass it to tool_call('name') before tc.end()"
            )

        self._ended = True

        payload = self._build_payload()
        _send_payload(payload, "tool_call", api_key=self._api_key, tcc_url=self._tcc_url)

    def _build_payload(self) -> Dict[str, Any]:
        end_time = _now_iso()

        payload: Dict[str, Any] = {
            "type": "tool_call",
            "run_id": self._run_id,
            "tool_call_id": self._tool_call_id,
            "tool_name": self._name or "unknown",
            "start_time": self._start_time,
            "end_time": end_time,
            "status_code": self._status_code,
        }

        if self._status_message is not None:
            payload["status_message"] = self._status_message
        if self._args is not None:
            payload["args"] = self._args
        if self._result is not None:
            payload["result"] = self._result

        return payload


def tool_call(
    run_id: str,
    tool_call_id: Optional[str] = None,
    tool_name: Optional[str] = None,
    api_key: Optional[str] = None,
    tcc_url: Optional[str] = None,
) -> ToolCall:
    return ToolCall(run_id=run_id, tool_call_id=tool_call_id, tool_name=tool_name, api_key=api_key, tcc_url=tcc_url)
